### ISSUE TITLE
MockBean does not resolve the right type for abstract test class

### ISSUE THREAD

**Issue by onacit** [spring-projects/spring-boot#20916]
Please forgive me if this kind of issue is not appropriate. I already [SO-ed](https://stackoverflow.com/q/61152887/330457).

Just in cases of issue titles, this issue might have a same issue with https://github.com/spring-projects/spring-boot/issues/15969. I'm not sure.

I have a tightly coupled abstract service class with generic types.

I have an abstract service class which autowires a specific type of repository.

```java
@NoRepositoryBean
interface SomeRepository<T> { }

abstract class SomeService<T extends SomeRepository<U>, U ...> {

    @Autowired
    private U repositoryInstance;
}
```

Now I'm trying to create an abstract test class for subclasses of the service class.

```java
@SpringBootTest
abstract class SomeServiceTest<T extends SomeService<U>, U extends SomeRepository<V>, V ...> {

    @Autowired
    private T serviceInstance;

    // personally expected
    // to mock the serviceInstance.repositoryInstance, and it doesn't.
    @MockBean
    private U repositoryInstance; // != serviceInstance.repositoryInstance();
}
```

But mocking the bean in a test class of actual service class works.
(Which is I intended that extended modules don't have to do.)

```java
class OtherServiceTest
        extends SomeServiceTest<OtherService, OtherRepository, ...> {

    @TestConfiguration
    OtherServiceTestConfiguration {

        // WORKS!!!
        // == serviceInstance().repositoryInstance();
        @MockBean private OtherRepository repositoryInstance;
    }
}

class AnotherServiceTest
        extends SomeServiceTest<AnotherService, AnotherRepository, ...> {

    // WORKS!!!
    // == serviceInstance().repositoryInstance();
    @MockBean private AnotherRepository repositoryInstance;
}
```

Is there any intrinsic mechanism for mocking the `SomeServiceTest#serviceIntance.repositoryInstance`?

**Comment by mbhave** [comment]
This might be happening because the mock is created for the `SomeRepository` type instead of the actual type which is `OtherRepository` or `AnotherRepository`.

@onacit Just to be sure, could you provide a minimal sample that we can run to reproduce the issue?

**Comment by onacit** [comment]
@mbhave Thank you. I managed to create a simplest and reproducible project from my original code base.

https://github.com/onacit/spring-boot-issue447

There is only one `@Test` method there in an abstract class and there are two actual test classes inherit.

* `...some0.SomeProductRepositoryServiceTest`
* `...some1.SomeProductRepositoryServiceTest`

**Comment by onacit** [comment]
I don't mean to bother you, but. Does that sample project work for you?

**Comment by onacit** [comment]
@mbhave Hi. I know the Hollywood Principle, but, do we have any progress here? Thank you.

**Comment by mbhave** [comment]
Thanks for the sample @onacit. I have taken a look and it seems like a bug to me. It is marked `for: team-attention` which means we will discuss it on one of the weekly team calls.

**Comment by Xyaren** [comment]
Could this be backported to 2.2.x ? 
We would really appreciate that, as we are still relying on Elasticsearch Jest, that is not included in 2.3.x anymore.

**Comment by wilkinsona** [comment]
@Xyaren The fix was made in 2.2.9.

**Comment by Xyaren** [comment]
I'm sorry, picket I got confused between two issues. Thanks for the quick response!
